tensornetwork/block_sparse/block_tensor.py

```diff
@@ -626,3 +627,256 @@ def _find_transposed_diagonal_sparse_blocks(
+    """
+    A list of list of `bool`.
+    The flows, in the current shape and index order as determined by `ChargeArray._order`.
+    Returns:
+      List of List of bool
+    """
+
+    return [[self._flows[n] for n in o] for o in self._order]
+
+  @property
+  def flat_charges(self):
+    return self._charges
+
+  @property
+  def flat_flows(self):
+    return list(self._flows)
+
+  @property
+  def flat_order(self) -> List:
+    """
+    The flattened `ChargeArray._oder`.
+    """
+    return flatten(self._order)
+
+  @property
+  def sparse_shape(self) -> Tuple:
+    """
+    The sparse shape of the tensor.
+    Returns:
+      Tuple: A tuple of `Index` objects.
+    """
+
+    indices = [
+        Index([self._charges[n]
+               for n in s], [self._flows[n]
+                             for n in s])
+        for s in self._order
+    ]
+    return tuple(indices)
+
+  def todense(self) -> np.ndarray:
+    """
+    Map the sparse tensor to dense storage.
+    
+    """
+    return np.reshape(self.data, self.shape)
+
+  def reshape(
+      self,
+      shape: Union[List[Index], Tuple[Index, ...], List[int], Tuple[int, ...]]
+  ) -> "ChargeArray":
+    """
+    Reshape `tensor` into `shape.
+    `BlockSparseTensor.reshape` works the same as the dense 
+    version, with the notable exception that the tensor can only be 
+    reshaped into a form compatible with its elementary shape. 
+    The elementary shape is the shape determined by BlockSparseTensor._charges.
+    For example, while the following reshaping is possible for regular 
+    dense numpy tensor,
+    ```
+    A = np.random.rand(6,6,6)
+    np.reshape(A, (2,3,6,6))
+    ```
+    the same code for BlockSparseTensor
+    ```
+    q1 = U1Charge(np.random.randint(0,10,6))
+    q2 = U1Charge(np.random.randint(0,10,6))
+    q3 = U1Charge(np.random.randint(0,10,6))
+    i1 = Index(charges=q1,flow=False)
+    i2 = Index(charges=q2,flow=True)
+    i3 = Index(charges=q3,flow=False)
+    A=BlockSparseTensor.randn(indices=[i1,i2,i3])
+    print(A.shape) #prints (6,6,6)
+    A.reshape((2,3,6,6)) #raises ValueError
+    ```
+    raises a `ValueError` since (2,3,6,6)
+    is incompatible with the elementary shape (6,6,6) of the tensor.
+    
+    Args:
+      tensor: A symmetric tensor.
+      shape: The new shape. Can either be a list of `Index` 
+        or a list of `int`.
+    Returns:
+      BlockSparseTensor: A new tensor reshaped into `shape`
+    """
+    new_shape = []
+    for s in shape:
+      if isinstance(s, Index):
+        new_shape.append(s.dim)
+      else:
+        new_shape.append(s)
+
+    # a few simple checks
+    if np.prod(new_shape) != np.prod(self.shape):
+      raise ValueError("A tensor with {} elements cannot be "
+                       "reshaped into a tensor with {} elements".format(
+                           np.prod(self.shape), np.prod(new_shape)))
+
+    flat_dims = flatten(
+        [[self._charges[n].dim for n in o] for o in self._order])
```

Comment:
I believe you can do `[self._charges[n].dim for n in o for o in self._order]`

Reply:
you're right, that's  much better (you need to change order of the loops though)